### ISSUE TITLE
Protocol: an "all positions" spelling for reads and writes

### ISSUE THREAD

**Issue by can-goodfire** [goodfire-ai/causalab#22]
## Context

The protocol refactor (#20) left a flagged gap in its spec questions (Q6): there is
no way to say *"read (or write) at every token position"*. Today a `pos` must name a
single index, a fixed span, or a prompt variable's window
(`docs/intervention_protocol.md` §2.3, lines 68–83). That is a whole parity axis
missing, and it is the concrete blocker for a logit-lens style (layer × position)
read grid: the layer axis is a sweep, but the position axis has no spelling.

Confirmed as wanted; this issue is the minimal implementation.

## What exists today

| piece | file | state |
|---|---|---|
| position spec | `causalab/protocol/schema.py:311` `PositionSpec`, parser at `:730` | exactly one of `index` / `span` / `variable` |
| sugar + canonical form | `causalab/protocol/canonical.py:208` `_canon_position_spec` | expands bare int → `{"index": n}` |
| resolution | `causalab/neural/pytorch_hooks/encoding.py:162` `resolve_position` | per-row → padded-frame indices, bounds-checked |
| ragged reads | `causalab/neural/pytorch_hooks/executor.py` `_gather` / `RaggedValue` | **already supported** (`{"variable": …}` is ragged) |
| ragged saves | `causalab/neural/pytorch_hooks/outputs.py:30` `TensorFile.add` | **already supported** (flat + `.widths`) |
| overlap rule | `causalab/protocol/validate.py:474` `_provably_disjoint` | conservative: not-provably-disjoint ⇒ overlapping |

So most of the machinery is in place. What is missing is the *spelling* plus its
resolution, and a handful of one-line integrations.

## Proposal

**Spelling.** A fourth mutually-exclusive anchor on a position spec, written inline
on the read or write — plus bare-string sugar, canonicalizing to the same thing:

```json
"reads": {"r": {"site": "resid", "pos": {"all": true}, "model": "original", "input": "base"}}
"reads": {"r": {"site": "resid", "pos": "all",         "model": "original", "input": "base"}}
```

`{"all": true}` carries no parameters, so unlike `{"variable": …}` or `{"span": …}` it
has nothing to gain from a named `positions` entry — and rule 11 (dead declarations)
would refuse an unreferenced one anyway. The sugar is the same shape as the existing
bare-int sugar; `all` joins `RESERVED_NAMES` (`causalab/protocol/schema.py:168`) so no
positions entry can shadow it.

*Alternative considered and rejected:* making `pos` omissible. `pos` is required
today, so omission would be a silent semantic change to any future document that
forgets it — explicit is better.

**Semantics.** `{"all": true}` resolves per row to **every real content token**:
`range(batch.content_start(row), batch.padded_len)` in the left-padded frame. That
excludes padding and excludes any chat prefix — consistent with `{"index": n}` for
`n ≥ 0`, which is already rebased past the prefix. `scope` / `relative_to` are
refused with `all` (they modify an index or span), matching the existing
`variable` rule.

**Rows are ragged.** Different rows have different content lengths, so an
all-positions read is a `RaggedValue` in general — a path the backend already
handles end to end (gather → featurize → save).

**Validation.** Rules 8/9 only compare *writes* (`_check_write_collisions`,
`causalab/protocol/validate.py:540`), so an all-positions *read* touches no overlap
rule at all — the "reads and edits need different overlap rules" worry does not
materialize. For writes, an `all` spec is never provably disjoint from anything, so
`_provably_disjoint` must return `False` whenever either side is `all`. That is
already the fallthrough, but it should be an explicit early return with a comment
and its own test rather than an accident of control flow.

**Digests are untouched.** `canonicalize` walks the *raw* tree
(`causalab/protocol/canonical.py:95`), so a new optional field adds nothing to the
canonical form of existing documents. `tests/protocol/corpus_digests.json` must not
change — that is an acceptance criterion, not a hope.

## Implementation plan

1. **`causalab/protocol/schema.py`** — add `all: Leaf | None = None` to
   `PositionSpec`; in `_parse_position_spec`, add `"all"` to the allowed keys and to
   the exactly-one-anchor check, accept the bare string `"all"` as sugar (alongside
   the bare int), refuse `scope`/`relative_to` combined with `all`, and refuse a
   non-`true` value. Add `"all"` to `RESERVED_NAMES`.
2. **`causalab/protocol/canonical.py`** — `_canon_position_spec` expands the bare
   string `"all"` → `{"all": true}`.
3. **`causalab/neural/pytorch_hooks/encoding.py`** — `resolve_position` returns
   `check(list(range(batch.content_start(row), batch.padded_len)))` for an `all`
   spec; extend the module docstring's position-rule list.
4. **`causalab/protocol/validate.py`** — explicit `if a.all or b.all: return False`
   at the top of `_provably_disjoint`, with the reasoning in a comment; note it in
   the module docstring's rule-8 interpretation bullet.
5. **`causalab/neural/pytorch_hooks/executor.py`** — the ragged-write refusal in
   `_address_writer` stays, but its message should name the case
   ("an all-positions write over rows of differing length"). No new write path.
6. **`docs/intervention_protocol.md`** — one row in the §2.3 table
   (`| {"all": true} | every content token of the row — ragged across rows |`), one
   row for the `"all"` string sugar, a sentence on the pad/chat-prefix exclusion,
   and `"all"` added to the reserved-name list in checklist rule 3 (line 321) and to
   the sugar list in the canonical-stamp principle (line 376).

## Tests

- `tests/neural/pytorch_hooks/test_positions_frame.py` — `all` on a two-row batch of
  unequal length: each row's indices are exactly its unmasked tokens, contiguous,
  ending at `padded_len - 1`; nothing under the left pad.
- `tests/neural/pytorch_hooks/test_read_oracle.py` — parity against a hand-hooked
  capture on tiny-random: an `embeddings` read at all positions equals the
  embedding rows for the unmasked tokens (the `{"variable": …}` oracle test at
  `:200` is the template). Include a document reading `lm_head` at all positions
  swept over layers, executed end to end, to pin the grid case the gap blocks.
- `tests/protocol/test_schema.py` — bare `"all"` sugar parses; two anchors refuse;
  `scope` + `all` refuses; a positions entry named `all` refuses.
- `tests/protocol/test_canonical.py` — `"all"` canonicalizes to `{"all": true}`;
  corpus digests unchanged.
- `tests/protocol/test_validation_rules.py` — one failing document: two absolute
  writes at one site, one at `{"index": -1}` and one at `{"all": true}`, refused by
  rule 8.

## Non-goals (named so they are not silently assumed)

- **Metrics over multi-position reads.** `compute_metric` reduces exactly one
  position per example and refuses more (`causalab/neural/pytorch_hooks/metrics.py:44`).
  So this issue unblocks the all-positions *read grid* (saved as a tensor), not a
  per-position *metric* grid — that needs a metric-domain change and a `position`
  column in the metric table, and belongs to the metrics-over-any-read work.
- **Ragged writes.** A write at `all` over rows of unequal length still refuses in
  the v1 reference backend. It works when every row has the same length. A masked
  write path (and the operand expansion it implies) is separate follow-up work.
- No new component vocabulary, no generation, no backend other than pytorch-hooks.

## Acceptance

- A document reading `lm_head` at all layers × all positions loads, validates and
  executes on tiny-random, and matches a hand-hooked oracle capture.
- `all` is refused where it is meaningless (with `scope`/`relative_to`, as a
  declared name, with a second anchor) and each refusal has a test.
- Rule 8 refuses an absolute write at `all` colliding with an absolute write at any
  other position; rule-9 dims disjointness behaves the same at an `all` address.
- `tests/protocol/corpus_digests.json` is byte-identical — the new field changes no
  existing document's canonical form.

Depends on #20 (the protocol refactor); the plan is written against that branch.
